File: Program1/src/List.java
//-----------------------------------------------------------------------------
//  List.java
//  Name: Yongsheng Cui
//  ID: 1491148
//  HW: PA1
//  This file will implement the List ADT operation
//-----------------------------------------------------------------------------
public class List {
	// private class Node
	private class Node {
		int data;
		Node prev;
		Node next;
		boolean eq = false;

		// Constructor - create data, next and prev in Node
		Node(int data) {
			this.data = data;
			this.prev = null;
			this.next = null;
		}

		// Returns String representation of data
		public String toString() {
			return String.valueOf(data);
		}

		// Returns true if the twoNode are equal
		public boolean equals(Object n) {
			boolean eq = false;
			if (n instanceof Node) {
				Node that = (Node) n;
				eq = (this.data == that.data);
			}
			return eq;
		}
	}

	// Private element
	private Node front;
	private Node back;
	private Node cursor;
	private int index;
	private int length;

	// Constructor
	//create a new empty list
	public List() {
		this.length = 0;
		this.index = -1;
		this.cursor = null;
		this.front = null;
		this.back = null;
	}

	// -- Access functions --------------------------------------
	// returns the number of elements in this List
	public int length() {
		return length;
	}

	// index()
	// if cursor is defined, returns the index of
	// the current element,
	// otherwise return -1
	public int index() {
		if (cursor == null) {
			return -1;
		}
		return index;
	}

	// returns front element. pre: length() > 0
	public int front() {
		if (length() > 0) {
			return front.data;
		} else
			throw new RuntimeException("Error: cannot called front() on empty list!");
	}

	// returns back element. pre: length() > 0
	public int back() {
		if (length() > 0) {
			return back.data;
		} else
			throw new RuntimeException("Error: cannot called back() on empty list!");
	}

	// returns the cursor element. pre: length() > 0 and index() > 0
	public int get() {
		if (length() <= 0) {
			throw new RuntimeException("Error: cannot called get() on empty list!");
		}
		if (cursor == null) {
			throw new RuntimeException("Error: cannot called get() if cursor is null");
		}
		return cursor.data;
	}

	// returns true if this List and L are the same integer
	// sequence. the cursor is ignored in both lists.
	public boolean equals(List L) {
		if (this.length == L.length) {
			Node FirstIterator = this.front;
			Node SecondIterator = L.front;
			while (FirstIterator != null && SecondIterator != null) {
				if (FirstIterator.data == SecondIterator.data) {
					FirstIterator = FirstIterator.next;
					SecondIterator = SecondIterator.next;
				} else return false;
			} return true;
		} else {
			return false;
		}
			
	}

	// -- Manipulation procedure --------------------------

	// rests this List to it original empty state
	public void clear() {

		this.length = 0;
		this.index = -1;
		this.cursor = null;
		this.front = null;
		this.back = null;

	}

	// if List is non-empty, places the cursor under
	// the front element, otherwise does nothing
	public void moveFront() {
		if (length() > 0) {
			cursor = front;
			index = 0;
		}
	}

	// if list is non-empty, places the cursor under
	// the back element, otherwise does nothing
	public void moveBack() {
		if (length() > 0) {
			cursor = back;
			index = length - 1;
		}
	}
	
	// if cursor is defined and not at front, moves cursor
	// one step toward front of this List, cursor
	// becomes undefined, if cursor is undefined does nothing
	public void movePrev() {
		if (cursor != front && index != -1) {
			cursor = cursor.prev;
			index--;
		} else if (cursor == front && index != -1) {
			cursor = null;
			index = -1;
		}
	}

	// if cursor is defined and not at front, move cursor
	// one step toward back of this List, if cursor is defined,
	// and at front, cursor becomes undefined, if
	// cursor is undefined, does nothing
	public void moveNext() {
		if (cursor != null && index == length - 1) {
			cursor = null;
			index = -1;
		} else if (cursor != null && index != length - 1) {
			cursor = cursor.next;
			index++;
		}
	}

	// insert new element into this List. if List is non-empty,
	// insertion takes place before front element
	public void prepend(int data) {
		Node newNode = new Node(data);
		if (front == null) {
			front = newNode;
			back = newNode;
			cursor = front;
		} else {
			front.prev = newNode;
			newNode.next = front;
			newNode.prev = null;
			front = newNode;
			index++;
		}
		length++;
	}

	// insert new element into this List. if this lis is non-empty
	// insertion takes place after back element
	public void append(int data) {
		Node newNode = new Node(data);
		if (back == null) {
			front = newNode;
			back = newNode;
			cursor = back;
		} else {
			back.next = newNode;
			newNode.prev = back;
			back = newNode;
			newNode.next = null;
		}
		length++;
	}

	// insert new element before cursor
	// pre: length() > 0, index() > 0
	public void insertBefore(int data) {
		Node newNode = new Node(data);

		if (length == 0) {
			throw new RuntimeException("Error: cannot called nsertBefore() on empty list!");

		}

		else {

			newNode.next = cursor;

			if (cursor.prev != null) {

				newNode.prev = cursor.prev;

				cursor.prev.next = newNode;

			}

			cursor.prev = newNode;

			if (newNode.prev == null)

				front = newNode;

		}

		index++;

		length++;

	}

	// insert new element after cursor
	// pre: length() > 0, index() > 0
	public void insertAfter(int data) {
		if (length() < 0 && index < 0)

			throw new RuntimeException("List error: insertAfter() called on empty List");

		if (cursor == back) {

			append(data);

		}

		else {

			Node temp = new Node(data);

			cursor.next.prev = temp;

			temp.next = cursor.next;

			temp.prev = cursor;

			cursor.next = temp;

			length++;

		}
	}

	// deletes front element. pre: length() > 0
	public void deleteFront() {
		if (length() < 1) {
			throw new RuntimeException("Error: cannot deleteFront() empty list.");
		} else {
			if (front == cursor) {
				cursor = null;
				index = -1;
			}
			if (length > 1) {
				front = front.next;
				front.prev = null;
				if (cursor != null) {
					index--;
				}
			}
			length--;
		}
	}

	// deletes back element. pre: length() > 0
	public void deleteBack() {
		if (length < 1) {
			throw new RuntimeException("Error :cannot deleteBack() on an empty list");
		}
		if (back == cursor) {
			cursor = null;
			index = -1;
		}
		back.next = null;
		back = back.prev;

		length--;
	}

	// deletes cursor element, making cursor undefined
	// pre: length() > 0, index() > 0
	public void delete() {
		if(length < 1)
		{
			throw new RuntimeException("List Class Exception: cannot delete() on empty list");
		}
		if(index() < 0)
		{
			throw new RuntimeException("List Class Exception: cannot delete() with undefined cursor");
		}
		if(cursor == front)
		{
			deleteFront();
		}
		else if(cursor == back)
		{
			deleteBack();
		}
		else
		{
			cursor.prev.next = cursor.next;
			cursor.next.prev = cursor.prev;
			cursor = null;			
			length--;
		}
		index = -1;
	}

	// -- Other methods --------------------------------------

	 // Overrides Object's toString method. Returns a String
	 // representation of this List consisting of a space
	// separated sequence of integers, with front on left.
	@Override
	public String toString() {
		Node N = front;
		String output = new String();
		while(N != null)
		{
			output = output + String.valueOf(N.data) + " ";
			N = N.next;
		}
		return output;
	}

	// returns a new List representing the same integer
	// sequence as this List. the cursor in the new list
	// is undefined, regardless of the state of the cursor
	// in this list. the List is unchanged
	public List copy() {
		List copyList = new List();
			Node N = front;
			while (N != null) {
				copyList.append(N.data);
				N = N.next;
			}
			copyList.cursor = null;
			copyList.index = -1;
		return copyList;
	}
}

File: pa2/List.c
/*
 * List.c
 *
 *  Created on: Oct 17, 2018
 *      Author: yongsheng cui
 */
#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// private NodeObj type
typedef struct NodeObj
{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
}  NodeObj;


// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj
{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
}  ListObj;

/*****CONSTRUCTORS - DESTRUCTORS*****/
// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data1, Node prev1, Node next1){
   Node N = malloc(sizeof(NodeObj));
   N->data = data1;
   N->prev = prev1;
   N->next = next1;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN) {
   if(pN != NULL && *pN != NULL) {
      free(*pN);
      *pN = NULL;
   }
}

// new List()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}

// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL.S
void freeList(List* pL) {
   if(pL != NULL && *pL != NULL) {
      Node temp = (*pL)->front;
      while(temp != NULL) {
         Node current = temp;
         temp = temp->next;
         free(current);
      }

      //deallocates the memory
      free(*pL);
      *pL = NULL;
   }
}

/*****ACCESS FUNCTIONS*****/

// Returns the length of L.
int length(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   return L->length;
}

//Returns the index of L
int index(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling index() on NULL List reference\n");
      exit(1);
   }
   return L->index;
}

// Returns front element. Pre: length()>0
int front(List L) {
   if(L == NULL)
    {
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: calling front() on empty List reference\n");
      exit(1);
   }
   return L->front->data;
}

// Returns back element. Pre: length()>0
int back(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: calling back() on empty List reference\n");
      exit(1);
   }
   return L->back->data;
}

// Returns cursor element. Pre: length()>0, index()>=0
int get(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling get() on NULL List reference\n");
      exit(1);
   }
   if(L->index < 0)
   {
      printf("List Error: calling get() with an undefined index on List\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: calling get() on empty List reference\n");
      exit(1);
   }
   return L->cursor->data;
}

// Returns true if this List and L are the same integer
// sequence. The cursor is ignored in both lists.
int equals(List A, List B)
{
   if(A == NULL || B == NULL)
    {
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   if(A->length != B->length)
   {
      return 0;
   }

   //set the current node in front to B->front
   Node currentFront = B->front;
   Node temp = A->front;

   while(currentFront->next != NULL && temp->next != NULL)
   {
      if(currentFront->data != temp->data)
         return 0;
      currentFront = currentFront->next;
      temp = temp->next;
   }
   return 1;
}
/*****ACCESS FUNCTIONS END*****/


/*****MANIPULATION FUNCTIONS*****/

// Resets this List to its original empty state.
void clear(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling clear() on NULL List reference\n");
      exit(1);
   }
   Node temp = L->front;
   while(temp != NULL)
   {
      Node current = temp;
      temp = temp->next;

      //calling free(Node)
      //deallocates the memory
      free(current);
   }

   //set front, back of List and cursor to NULL
   L->front = NULL;
   L->back = NULL;
   L->cursor = NULL;

   //set the length back to 0
   L->length = 0;
   //set the index back to -1
   L->index = -1;
}

// If List is non-empty, places the cursor under the front element,
// otherwise does nothing.
void moveFront(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling moveFront() on NULL List reference\n");
      exit(1);
   }
   if(L->length > 0)
   {
      L->cursor = L->front;
      L->index = 0;
   }
}

// If List is non-empty, places the cursor under the back element,
// otherwise does nothing.
void moveBack(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling moveBack() on NULL List reference\n");
      exit(1);
   }
   if(L->length > 0)
   {
      L->cursor = L->back;
      L->index = L->length - 1;
   }
}

// If cursor is defined and not at front, moves cursor one step toward
// front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing.
void movePrev(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling movePrev() on NULL List reference\n");
      exit(1);
   }
   if(L->cursor != NULL && L->index != 0)
   {
      L->cursor = L->cursor->prev;
      --L->index;
   }
   else if(L->cursor != NULL && L->index == 0)
   {
      L->cursor = NULL;
      L->index = -1;
   }
}

// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
void moveNext(List L)
{
   if(L == NULL)
    {
      printf("List Error: calling moveNext() on NULL List reference\n");
      exit(1);
   }
   if(L->cursor != NULL && L->index != L->length - 1)
   {
      L->cursor = L->cursor->next;
      ++L->index;
   }
   else if(L->cursor != NULL && L->index == L->length - 1)
   {
      L->cursor = NULL;
      L->index = -1;
   }
}

// Insert new element into this List. If List is non-empty,
// insertion takes place before front element.
void prepend(List L, int data)
{
   if(L == NULL)
    {
      printf("List Error: calling prepend() on NULL List reference\n");
      exit(1);
   }
   Node temp = newNode(data, NULL, L->front);
   if(L->front == NULL) L->back = temp;
   else
    {
        L->front->prev = temp;
    }
   L->front = temp;
   ++L->length;
}

void append(List L, int data)
{
   if(L == NULL)
    {
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   Node temp = newNode(data, L->back, NULL);
   if(L->front == NULL) L->front = temp;
   else
   {
       L->back->next = temp;
   }
   L->back = temp;
   ++L->length;
}

// Inserts new element after cursor.
// Pre: length()>0, index()>=0
void insertAfter(List L, int data)
{
   if(L == NULL)
    {
      printf("List Error: calling insertAfter() on NULL List reference\n");
      exit(1);
   }
   if(L->index < 0)
   {
      printf("List Error: insertAfter() called with an undefined index on List");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: insertAfter() called on an empty List");
      exit(1);
   }
   Node temp = newNode(data, L->cursor, L->cursor->next);

   if(L->cursor->next != NULL) L->cursor->next->prev = temp;

   else
      L->back = temp;
   L->cursor->next = temp;
   ++L->length;
}

// Insert new element before cursor.
// Pre: length() > 0, index() >= 0
void insertBefore(List L, int data)
{
   if(L == NULL)
    {
      printf("List Error: calling insertBefore() on NULL List reference\n");
      exit(1);
   }
   if(L->index < 0)
   {
      printf("List Error: insertBefore() called with an undefined index on List");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: insertBefore() called on an empty List");
      exit(1);
   }
   Node temp = newNode(data, L->cursor->prev, L->cursor);
   if(L->cursor->prev != NULL) L->cursor->prev->next = temp;

   else
      L->front = temp;
   L->cursor->prev = temp;
   ++L->length;
}

// Deletes the front element.
// Pre: length() > 0
void deleteFront(List L)
{
   if(L == NULL)
    {
      printf("List Error: deleteFront() called on NULL List reference\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: deleteFront() called on an empty List\n");
      exit(1);
   }
   if(L->cursor == L->front)
   {
      L->cursor = NULL;
      L->index = -1;
   }

   Node temp = L->front;
   L->front = L->front->next;
   L->front->prev = NULL;
   --L->length;
   freeNode(&temp);
}

// Deletes the back element.
// Pre: length() > 0
void deleteBack(List L)
{
   if(L == NULL)
    {
      printf("List Error: deleteBack() called on NULL List reference\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: deleteBack() called on an empty List\n");
      exit(1);
   }
   if(L->cursor == L->back)
   {
      L->cursor = NULL;
      --L->index;
   }

   Node temp = L->back;
   L->back = L->back->prev;
   L->back->next = NULL;
   --L->length;
   freeNode(&temp);
}

// Deletes cursor element, making cursor undefined.
// Pre: length() > 0, index() >= 0
void delete(List L)
{
   if(L == NULL)
    {
      printf("List Error: delete() called on NULL List reference\n");
      exit(1);
   }
   if(L->length < 1)
   {
      printf("List Error: delete() called with an undefined index on List\n");
      exit(1);
   }
   if(L->index < 0)
   {
      printf("List Error: delete() called on empty List");
      exit(1);
   }
   if(L->cursor == L->back)
   {
      deleteBack(L);
   }
   else if(L->cursor == L->front)
   {
      deleteFront(L);
   }
   else
   {
      Node temp = L->cursor;
      L->cursor->prev->next = L->cursor->next;
      L->cursor->next->prev = L->cursor->prev;
      freeNode(&temp);
      L->cursor = NULL;
      L->index = -1;
      --L->length;
   }
}

/*****MANIPULATION FUNCTIONS*****/

// Overrides Object's toString method. Returns a String
// representation of this List consisting of a space
// separated sequence of integers, with front on left.
void printList(FILE* out, List L)
{
   Node temp = L->front;
   while(temp != NULL)
    {
      printf("%d ", temp->data);
      temp = temp->next;
   }
}

// Returns a new List representing the same integer sequence as this
// List. The curor in the new list is undefined, regardless of the
// state of the cursor in this List. This List is unchanged.
List copyList(List L)
{
   List c = newList();
   Node temp = L->front;
   while(temp != NULL)
    {
      append(c, temp->data);
      temp = temp->next;
   }
   return c;
}

